drivers/adc/adc_stm32.c

```diff
@@ -0,0 +1,588 @@
+#else
+	LL_ADC_SetChannelSamplingTime(adc,
+		__LL_ADC_DECIMAL_NB_TO_CHANNEL(id),
+		table_samp_time[acq_time_index]);
+#endif
+}
+
+static int adc_stm32_channel_setup(struct device *dev,
+			    const struct adc_channel_cfg *channel_cfg)
+{
+#if defined(CONFIG_SOC_SERIES_STM32F0X) || defined(CONFIG_SOC_SERIES_STM32L0X)
+	struct adc_stm32_data *data = dev->driver_data;
+#endif
+	int acq_time_index;
+
+	if (channel_cfg->channel_id >= STM32_CHANNEL_COUNT) {
+		LOG_ERR("Channel %d is not valid", channel_cfg->channel_id);
+		return -EINVAL;
+	}
+
+	acq_time_index = adc_stm32_check_acq_time(
+				channel_cfg->acquisition_time);
+	if (acq_time_index < 0) {
+		return acq_time_index;
+	}
+#if defined(CONFIG_SOC_SERIES_STM32F0X) || defined(CONFIG_SOC_SERIES_STM32L0X)
+	if (data->acq_time_index == -1) {
+		data->acq_time_index = acq_time_index;
+	} else {
+		/* All channels of F0/L0 must have identical acquisition time.*/
+		if (acq_time_index != data->acq_time_index) {
+			return -EINVAL;
+		}
+	}
+#endif
+
+	if (channel_cfg->differential) {
+		LOG_ERR("Differential channels are not supported");
+		return -EINVAL;
+	}
+
+	if (channel_cfg->gain != ADC_GAIN_1) {
+		LOG_ERR("Invalid channel gain");
+		return -EINVAL;
+	}
+
+	if (channel_cfg->reference != ADC_REF_INTERNAL) {
+		LOG_ERR("Invalid channel reference");
+		return -EINVAL;
+	}
+
+	adc_stm32_setup_speed(dev, channel_cfg->channel_id,
+				  acq_time_index);
+
+	LOG_DBG("Channel setup succeeded!");
+
+	return 0;
+}
+
+#if !defined(CONFIG_SOC_SERIES_STM32F2X) && \
+	!defined(CONFIG_SOC_SERIES_STM32F4X) && \
+	!defined(CONFIG_SOC_SERIES_STM32F7X) && \
+	!defined(CONFIG_SOC_SERIES_STM32F1X)
+static void adc_stm32_calib(struct device *dev)
+{
+	struct adc_stm32_cfg *config =
+		(struct adc_stm32_cfg *)dev->config->config_info;
+	ADC_TypeDef *adc = config->base;
+
+#if defined(CONFIG_SOC_SERIES_STM32F3X) || \
+	defined(CONFIG_SOC_SERIES_STM32L4X)
+	LL_ADC_StartCalibration(adc, LL_ADC_SINGLE_ENDED);
+#elif defined(CONFIG_SOC_SERIES_STM32F0X) || \
+	defined(CONFIG_SOC_SERIES_STM32L0X)
+	LL_ADC_StartCalibration(adc);
+#endif
+	while (LL_ADC_IsCalibrationOnGoing(adc))
+		;
+}
+#endif
+
+static int adc_stm32_init(struct device *dev)
+{
+	struct adc_stm32_data *data = dev->driver_data;
+	const struct adc_stm32_cfg *config = dev->config->config_info;
+	struct device *clk =
+		device_get_binding(STM32_CLOCK_CONTROL_NAME);
+	ADC_TypeDef *adc = (ADC_TypeDef *)config->base;
+
+	LOG_DBG("Initializing....");
+
+	data->dev = dev;
+#if defined(CONFIG_SOC_SERIES_STM32F0X) || defined(CONFIG_SOC_SERIES_STM32L0X)
+	data->acq_time_index = -1;
+#endif
+
+	if (clock_control_on(clk,
+		(clock_control_subsys_t *) &config->pclken) != 0) {
+		return -EIO;
+	}
+
+#if defined(CONFIG_SOC_SERIES_STM32L4X)
+	LL_ADC_DisableDeepPowerDown(adc);
+	LL_ADC_EnableInternalRegulator(adc);
+	u32_t wait_loop_index = ((LL_ADC_DELAY_INTERNAL_REGUL_STAB_US / 10UL) *
+				(SystemCoreClock / (100000UL * 2UL)));
+	while (wait_loop_index != 0UL) {
```

Comment:
k_busy_wait(LL_ADC_DELAY_INTERNAL_REGUL_STAB_US);

Reply:
That's better.